### PR TITLE
Mtp optimizations round1

### PR DESCRIPTION
**Pull request by ameynaik-hub**
# Deepseek MTP optimization.

Optimize mtp head for deepseek

1. eh_proj gemm sharding
2. rmsnorm cuda stream parallelization. 
3. lm_head top1 using local argmax --> AG --> argmax
4. piecewise torch.compile to fuse / opt multiple native kernels. 

## Test Coverage

BS1 deepseek perf improves from 411-->431 TPS on B200x8 TP8EP1.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Summary by CodeRabbit

* **New Features**
  * Improved performance and efficiency for token state extraction and normalization in DeepseekV3 models.
  * Enhanced support for distributed and parallel token selection in speculative decoding with tensor parallelism.
  * Added new helper functions for optimized token sampling and draft token management.

* **Refactor**
  * Modularized and compiled key tensor operations for better clarity and speed.
  * Updated constructors and initialization logic to consistently use model configuration context across relevant components.

* **Chores**
  * Adjusted internal interfaces to support expanded configuration and parallelism features.

### CODE REVIEW

**Review by Kefeng-Duan — APPROVED** [review]

**Review by hello-11 — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py [hunk 1]
```diff
@@ -130,28 +130,38 @@ class DeepseekV3MTPHead(nn.Module):
     def __init__(self, model_config: ModelConfig[PretrainedConfig]):
         super().__init__()
         config = model_config.pretrained_config
+        self.model_config = model_config
 
         self.norm = RMSNorm(hidden_size=config.hidden_size,
                             eps=config.rms_norm_eps,
                             dtype=config.torch_dtype)
 
+    @torch.compile(mode="max-autotune-no-cudagraphs")
```
[the remark this answers is not in the document]
Comment (dc3671) [reply tensorrt_llm/_torch/models/modeling_deepseekv3.py:139]: @ameynaik-hub I'll recommend using `@torch.compile(options={"max-autotune": True})`. In some versions of pytorch, using preset `mode` will cause compile error. See this commit  https://github.com/NVIDIA/TensorRT-LLM/pull/5865/files

**Review by QiJune — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ameynaik-hub) [reply tensorrt_llm/_torch/models/modeling_deepseekv3.py]: Thanks @dc3671 just updated and verified with the above mentioned change.

**Comment by QiJune** [comment]
/bot skip --comment "reuse previous passed CI"
